checker-qual/src/main/java/org/checkerframework/checker/calledmethods/qual/EnsuresCalledMethodsOnException.java

```diff
@@ -0,0 +1,83 @@
+package org.checkerframework.checker.calledmethods.qual;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.checkerframework.framework.qual.InheritedAnnotation;
+
+/**
+ * Indicates that the method, if it terminates by throwing an exception, always invokes the given
+ * methods on the given expressions. This annotation is repeatable, which means that users can write
+ * more than one instance of it on the same method (users should NOT manually write an
+ * {@code @EnsuresCalledMethodsOnException.List} annotation, which the checker will create from
+ * multiple copies of this annotation automatically).
+ *
+ * <p>Consider the following method:
+ *
+ * <pre>
+ * &#64;EnsuresCalledMethodsOnException(value = "#1", methods = "m")
+ * public void callM(T t) { ... }
+ * </pre>
+ *
+ * <p>This method guarantees that {@code t.m()} is always called before the method throws an
+ * exception.
+ *
+ * @see EnsuresCalledMethods
+ * @checker_framework.manual #called-methods-checker Called Methods Checker
+ */
+@Target({ElementType.METHOD, ElementType.CONSTRUCTOR})
+@Repeatable(EnsuresCalledMethodsOnException.List.class)
+@Retention(RetentionPolicy.RUNTIME)
+@InheritedAnnotation
+public @interface EnsuresCalledMethodsOnException {
+
+  /**
+   * Returns Java expressions that have had the given methods called on them after the method throws
+   * an exception.
+   *
+   * @return an array of Java expressions
+   * @checker_framework.manual #java-expressions-as-arguments Syntax of Java expressions
+   */
+  String[] value();
+
+  // NOTE 2023/10/6: There seems to be a fundamental limitation in the dataflow framework that
+  // prevent this feature.  Specifically, every method has a SINGLE exceptional exit block, meaning
+  // all information about what happens down different exception paths gets totally erased.
```

Comment:
Here's (a simplified version of) the test case that inspired the comment:

```java
  class E1 extends Exception {}
  class E2 extends Exception {}

  abstract void m1() throws E1;
  abstract void m2() throws E2;

  void m3() throws Exception {
    try {
      m1();
    } finally {
      m2();
    }
  }
```

The method `m3` has a possible behavior where
 1. `m1()` throws `E1`
 2. the finally block runs `m2()` and it exits normally
 3. `m3` exits throwing `E1`

However, the CFG does not have an edge to "exceptional-exit" with the label `E1`:
[OwnedGenericValue-m3-ResourceLeak.pdf](https://github.com/typetools/checker-framework/files/13242674/OwnedGenericValue-m3-ResourceLeak.pdf)

Instead, there is a single edge from "marker (end of finally block for Throwable #0)" to "exceptional-exit" with the vague label `Throwable` (in the bottom-right of the CFG diagram).

Suppose `m1` had a post-condition that promised some behavior when it throws `E1`. I think it would not be possible to prove that `m3` has similar behavior when it throws `E1`, because once the finally block starts the dataflow framework "forgets" the details of what exception is being thrown. The "forgetting" happens near the top of the CFG, where all possible exceptions from `m1()` flow into the same successor (namely the start of the finally block).

Reply:
Rereading it now, I think the comment is mostly wrong, even if the conclusion is right. I will re-word it.

Reply:
Thanks, the example CFG shows the issue here.  I agree a re-wording would be great.